### PR TITLE
pb-4449: Fixed the issue of namespaceMapping getting overwritten in the volume.

### PR DESCRIPTION
**Pull request by siva-portworx**
**What type of PR is this?**
>bug

**What this PR does / why we need it**:
```
pb-4449: Fixed the issue of namespaceMapping getting overwritten in the volume
    stage update.
```

**Does this PR change a user-facing CRD or CLI?**:
No

**Is a release note needed?**:
```release-note
Issue: If the namespace mapping is not mentioned in the applicationrestore CR, the restore succeeds but volumes/resources are not getting restored.
User Impact: The user is forced to specify the namespacemapping, even if they was want to use source namespace by default.
Resolution: Fixed the issue of namespaceMapping getting overwritten in the volume restore stage.

```

**Does this change need to be cherry-picked to a release branch?**:
23.8 branch

Tested:
```
[root@ip-10-13-196-168 ~]# cat ar.yaml
apiVersion: stork.libopenstorage.org/v1alpha1
kind: ApplicationRestore
metadata:
  name: mysql
  namespace: mysql
spec:
  backupName: mysql
  backupLocation: bl
[root@ip-10-13-196-168 ~]#
```

```
[root@ip-10-13-196-168 ~]# kubectl describe applicationrestore -A
Name:         mysql
Namespace:    mysql
Labels:       <none>
Annotations:  <none>
API Version:  stork.libopenstorage.org/v1alpha1
Kind:         ApplicationRestore
Metadata:
  Creation Timestamp:  2023-09-20T11:33:40Z
  Finalizers:
    stork.libopenstorage.org/finalizer-cleanup
  Generation:  9
  Managed Fields:
    API Version:  stork.libopenstorage.org/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:kubectl.kubernetes.io/last-applied-configuration:
      f:spec:
        .:
        f:backupLocation:
        f:backupName:
    Manager:      kubectl-client-side-apply
    Operation:    Update
    Time:         2023-09-20T11:33:40Z
    API Version:  stork.libopenstorage.org/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"stork.libopenstorage.org/finalizer-cleanup":
      f:spec:
        f:includeOptionalResourceTypes:
        f:includeResources:
        f:namespaceMapping:
          .:
          f:mysql:
        f:rancherProjectMapping:
        f:replacePolicy:
        f:storageClassMapping:
      f:status:
        .:
        f:finishTimestamp:
        f:largeResourceEnabled:
        f:lastUpdateTimestamp:
        f:reason:
        f:resourceCount:
        f:resourcerestorestate:
        f:resources:
        f:restoredresourceCount:
        f:stage:
        f:status:
        f:totalSize:
        f:volumes:
    Manager:         stork
    Operation:       Update
    Time:            2023-09-20T11:34:04Z
  Resource Version:  1620201
  UID:               3e53cc6c-cdb8-4976-8e4f-d141bfae6ae3
Spec:
  Backup Location:                  bl
  Backup Name:                      mysql
  Include Optional Resource Types:  <nil>
  Include Resources:                <nil>
  Namespace Mapping:
    Mysql:                  mysql
  Rancher Project Mapping:  <nil>
  Replace Policy:
  Storage Class Mapping:    <nil>
Status:
  Finish Timestamp:        2023-09-20T11:34:04Z
  Large Resource Enabled:  false
  Last Update Timestamp:   2023-09-20T11:34:04Z
  Reason:                  Volumes and resources were restored up successfully
  Resource Count:          3
  Resourcerestorestate:
  Resources:
    Group:
    Kind:                  PersistentVolume
    Name:                  pvc-31cc01b5-a94c-43e7-89e8-c18ff3fa479a
    Namespace:
    Reason:                Resource restored successfully
    Status:                Successful
    Version:               v1
    Group:
    Kind:                  PersistentVolumeClaim
    Name:                  mysql-data
    Namespace:             mysql
    Reason:                Resource restored successfully
    Status:                Successful
    Version:               v1
    Group:                 apps
    Kind:                  Deployment
    Name:                  mysql
    Namespace:             mysql
    Reason:                Resource restored successfully
    Status:                Successful
    Version:               v1
  Restoredresource Count:  3
  Stage:                   Final
  Status:                  Successful
  Total Size:              45326336
  Volumes:
    Driver Name:  pxd
    Options:
    Persistent Volume Claim:      mysql-data
    Persistent Volume Claim UID:  aa56c5cb-c2ea-433a-bc11-45e8ccff4bb9
    Reason:                       Restore successful for volume
    Restore Volume:               pvc-31cc01b5-a94c-43e7-89e8-c18ff3fa479a
    Source Namespace:             mysql
    Source Volume:                pvc-aa56c5cb-c2ea-433a-bc11-45e8ccff4bb9
    Status:                       Successful
    Total Size:                   45326336
    Zones:                        <nil>
Events:                           <none>
```

### CODE REVIEW

**Review by prashanthpx — APPROVED** [review]

File: pkg/applicationmanager/controllers/applicationrestore.go [hunk 1]
```diff
@@ -780,7 +787,7 @@ func (a *ApplicationRestoreController) restoreVolumes(restore *storkapi.Applicat
 							v1.EventTypeWarning,
 							string(storkapi.ApplicationRestoreStatusFailed),
 							message)
-						_, err = a.updateRestoreCRInVolumeStage(namespacedName, storkapi.ApplicationRestoreStatusFailed, storkapi.ApplicationRestoreStageFinal, message, nil)
+						_, err = a.updateRestoreCRInVolumeStage(namespacedName, storkapi.ApplicationRestoreStatusFailed, storkapi.ApplicationRestoreStageFinal, message, nil, nil)
```
Comment (diptiranjanpx) [inline pkg/applicationmanager/controllers/applicationrestore.go:790]: why we are passing explicitly "nil" here instead of restore.Spec.NamespaceMapping?

**Review by diptiranjanpx — COMMENTED** [review]
Also, I hope no change is required when using nfs as backuplocation.

File: pkg/applicationmanager/controllers/applicationrestore.go (hunk 1 above)
Comment (siva-portworx) [reply pkg/applicationmanager/controllers/applicationrestore.go:790]: First time, when we enter restoreVolume, we will be persisting the namespacemapping to the CR at line 554. In other places, It is not necessary as it is already persisted to the CR, we get it when read the CR content in updateRestoreCRInVolumeStage()

**Comment by siva-portworx** [comment]
> Also, I hope no change is required when using nfs as backuplocation.

No, fix is not depend on the backuplocation type. Also the change is in common path in restoreVolume()

**Review by pallav-px — APPROVED** [review]

**Pull request merged**